chatgpt_wrapper/chatgpt.py

```diff
@@ -333,48 +411,222 @@ def ask_stream(self, prompt: str):
+            response=response.replace('\u200b','')
+
+            if len(response) > 0:
+                chunk = response[len(last_event_msg):]
+                last_event_msg = response
+                print(chunk,end='',flush=True)
+
+                if 'An error occurred. If this issue persists please contact us through' in response or\
+                    'The server had an error while processing your request.' in response:
+                    print("Error detected when receiving response")
+                    self.log.error("ChatGPT gave an Error (Retrying...): %s",response)
+                    # TODO: Choose the right conversation
+                    self.page.reload()
+                    time.sleep(10)
+                    for i in self.ask_stream_clicking(prompt,remaining_retry=remaining_retry-1):
+                        yield i
+                    return
+
+                self.log.info("< (ChatGPT) %s",chunk)
+                yield chunk
+
+            # if we saw the eof signal, this was the last event we
+            # should process and we are done
+            if len(eof_datas) > 0:
+                break
+
+            time.sleep(0.2)
+        else:
+            # Timeout
+            self.log.error("Timed out waiting for ChatGPT response. ")
+            self.page.reload()
+            time.sleep(10)
+            for i in self.ask_stream_clicking(prompt,remaining_retry=remaining_retry-1):
+                yield i
+            return
 
         self._cleanup_divs()
         self._gen_title()
 
-    def ask(self, message: str) -> str:
+
+    def ask(self, message: str, remaining_retry=2) -> str:
         """
         Send a message to chatGPT and return the response.
 
         Args:
             message (str): The message to send.
+            remaining_retry (int): How many retries left
 
         Returns:
             str: The response received from OpenAI.
         """
-        response = list(self.ask_stream(message))
-        return (
-            reduce(operator.add, response)
-            if len(response) > 0
-            else "Unusable response produced, maybe login session expired. Try 'pkill firefox' and 'chatgpt install'"
-        )
+        if remaining_retry==0:
+            self.log.error("Cannot ask ChatGPT. Repeatedly receiving empty response. ")
+            raise error.ChatGPTResponseError("Cannot ask ChatGPT. Repeatedly receiving empty response. ")
+
+        response = [i for i in self.ask_stream(message)]
+        if len(response)!=0:
+            return ''.join(response)
+        # Else: retry
+        self.log.error("Received empty response from ChatGPT. Retrying... ")
+        self.refresh_session()
+        return self.ask(message,remaining_retry=remaining_retry-1)
 
     def new_conversation(self):
         self.parent_message_id = str(uuid.uuid4())
         self.conversation_id = None
         self.conversation_title_set = None
+
+
+
+def main():
```

Comment:
Why is this in here? We already have code to interact at shell.

Reply:
Ah. It was used as more or less a unit test. Teach me if it doesn't belong here.

Reply:
Or I should remove this. 

Actually I had some hard time figuring out how to test this code. Now I found that I should use `pip install -e`. lol

:laughing:

Reply:
Removed.